tests/virt_autotest/xen_guest_irqbalance.pm

```diff
@@ -0,0 +1,192 @@
+use virt_utils qw(upload_virt_logs remove_vm restore_downloaded_guests);
+
+sub run_test {
+    my $self = shift;
+
+    return if is_kvm_host;
+    #save original guest configuration file in case of restore in post_fail_hook()
+    my $downloaded_xml_dir = "/tmp/download_vm_xml";
+    foreach my $guest (keys %virt_autotest::common::guests) {
+        unless (script_run("ls $downloaded_xml_dir/$guest.xml") == 0) {
+            assert_script_run "mkdir -p $downloaded_xml_dir";
+            assert_script_run "virsh dumpxml --inactive $guest > $downloaded_xml_dir/$guest.xml";
+        }
+    }
+
+    foreach my $guest (keys %virt_autotest::common::guests) {
+
+        #irqbalance test run on sle12sp3+ guest
+        next if guest_is_sle($guest, '<=12-sp2');
+
+        record_info("Test $guest");
+        prepare_guest_for_irqbalance($guest);
+
+        # with irqbalance.service is enabled, SMP affinity will be distributed among CPUs automatically even if IRQs are bound to specified CPUs in SLE15SP4
+        assert_script_run "ssh root\@$guest \"systemctl stop irqbalance.service\"" if script_run("ssh root\@$guest \"systemctl status irqbalance\"") eq 0;
+
+        my $nproc = script_output("ssh root\@$guest 'nproc'");
+
+        # Bind all NIC IRQs to cpu0 in guest
+        my @nic_irqs_id = split('\n', script_output("ssh root\@$guest \"grep vif /proc/interrupts | cut -d ':' -f1 | sed 's/^ *//'\""));
+        foreach my $irq_id (@nic_irqs_id) {
+            assert_script_run("ssh root\@$guest \"echo 1 > /proc/irq/$irq_id/smp_affinity\"");
+        }
+        my @affinities_with_binding_cpu0 = get_irq_affinities_from_guest($guest, @nic_irqs_id);
+        foreach (@affinities_with_binding_cpu0) {
+            die "Failed to bind NIC IRQs to CPU0" if $_ ne 1;
+        }
+
+        # Start the irqbalance.service again
+        assert_script_run("ssh root\@$guest \"systemctl start irqbalance\"");
+
+        # Check if the SMP affinities are distributed among CPUs
+        my @affinities_with_irqbalance = get_irq_affinities_from_guest($guest, @nic_irqs_id);
+        # for SLE15SP3 and prior, IRQs are distributed on one cpu core ramdomly.
+        # So it is correct if some of them may happen to run on cpu0
+        # but it would not be correct if all IRQs were bound to cpu0.
+        if (guest_is_sle($guest, '<=15-sp3')) {
+            my $multiply = 1;
+            $multiply *= $_ foreach (@affinities_with_irqbalance);
+	    print "julie: \$multiply=$multiply of affinities for $guest after irqbalance.\n";
+            die "SMP affinities were all bound to CPU0 with irqbalance.service enabled" if $multiply == 1;
+        }
+        # for SLE15SP4 (or newer release), IRQs are specified on all CPU cores
+        # fg. affinity value is '0xf' for machine with 4 CPU cores
+        else {
+            foreach (@affinities_with_irqbalance) {
+                die "It is not correct that one of the NIC IRQs was bound to cpu0 with irqbalance enabled." if $_ eq 1;
```

Comment:
Why checking $_ eq 1 ? It seems that any values other than 0xf are invalid. You'd better check $_ ne 0xf ?

Reply:
vaule=1 means the IRQ is still bound to CPU0, the irqbalance does not work as expected. The judgement is enough.

the balancing rule varies in different SLES. As far as I know, it is usually 'f' in 15SP4, but it can be '1', '2', '00001000' or other values in 15SP3/12SP5. I am afraid that the value is not definitly be 'f' in future release.

Reply:
If irqbalance works, I think the affinity value should not have 0 in place where is a corresponding cpu. Affinity value should distributed across all cpus. But you also mentioned sles of certan version does not give it a "f" even with irqbalance in place.

Reply:
yes, affinity distribute to all cpu in sle15sp4, ie. value 'f' which means "cpu0 cpu1 cup2 cup3'. but in older release, the affinity picks up one cpu ramdomly(appears to be), so the value may be '1'(cpu0), '2'(cpu1), '4'(cpu3), '000200'(cpu10)
'

Reply:
@Julie-CAO  Sorry for the late comment, I just managed to open github. According to some online material, the relationship between the affinity value and the cpu number is that if we turn the affinity value into binary number, which bit has 1 means which cpu has irq balanced. So if in 15sp4, all cpus are bound, then the affinity value should equal to nproc bits of 1 in binary way. We can use more precise checking here. HDYT?

Reply:
the online document I have found are all talking about the 15sp4 situations. if 15sp5/6 or sle-next apply to this rule as well, '1111...' would be the best checking value. We can make this as the convence since you and Wayne both think it is worthy. Let me check '1111...' for 15sp4+ guest, although it is not too easy.

Reply:
Compare to the default_smp_affinity instead of 1 for 15sp4+ guest, which is not perfect too, but it is better.